urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlab/examples/ConsecutiveIsomorphismSpecification.java

```diff
@@ -0,0 +1,350 @@
+            }
+            HashSet<Permutation> sp = ConsecutiveOrder.parents.get(p);
+            HashSet<Permutation> isp = apply(sp);
+            if (isp != null && ConsecutiveOrder.children.containsKey(isp)) {
+                HashSet<Permutation> possibleImages = ConsecutiveOrder.children.get(isp);
+                ArrayList<Permutation> pi = new ArrayList<Permutation>();
+                for (Permutation q : possibleImages) {
+                    if (!result.inRange(q)) {
+                        pi.add(q);
+                        break;
+                    }
+                }
+                if (!pi.isEmpty()) {
+                    result.add(p, pi.get(0));
+                }
+            }
+
+        }
+        result.fixBasis(n);
+        return result;
+    }
+
+    public ConsecutiveIsomorphismSpecification randomExtension(int n) {
+        ConsecutiveOrder.buildMaps(n);
+        ConsecutiveIsomorphismSpecification result = this.copy();
+        PermProperty inClass = new ConsecutiveAvoidanceTest(basis);
+        for (Permutation p : new Permutations(n)) {
+            if (!inClass.isSatisfiedBy(p)) {
+                continue;
+            }
+            HashSet<Permutation> sp = ConsecutiveOrder.parents.get(p);
+            HashSet<Permutation> isp = apply(sp);
+            if (isp != null && ConsecutiveOrder.children.containsKey(isp)) {
+                HashSet<Permutation> possibleImages = ConsecutiveOrder.children.get(isp);
+                ArrayList<Permutation> pi = new ArrayList<Permutation>();
+                for (Permutation q : possibleImages) {
+                    if (!result.inRange(q)) {
+                        pi.add(q);
+                    }
+                }
+                if (!pi.isEmpty()) {
+                    result.add(p, pi.get((int) (Math.random() * pi.size())));
+                }
+            }
+
+        }
+        result.fixBasis(n);
+        return result;
+    }
+
+    private HashSet<HashMap<Permutation, Permutation>> matchings(HashSet<Permutation> a, HashSet<Permutation> b) {
+        HashSet<HashMap<Permutation, Permutation>> result = new HashSet<HashMap<Permutation, Permutation>>();
+        Permutation[] domain = new Permutation[a.size()];
+        a.toArray(domain);
+        Permutation[] codomain = new Permutation[b.size()];
+        b.toArray(codomain);
+        if (domain.length <= codomain.length) {
+            for (int[] c : new Combinations(codomain.length, domain.length)) {
+                for (Permutation p : new Permutations(domain.length)) {
+                    HashMap<Permutation, Permutation> f = new HashMap<Permutation, Permutation>();
+                    for (int i = 0; i < domain.length; i++) {
+//                        System.out.println("Putting: " + domain[i] + " " + codomain[c[p.at(i)]]);
+                        f.put(domain[i], codomain[c[p.at(i)]]);
+                    }
+                    result.add(f);
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 245:254, 257:266.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.